### ISSUE TITLE
Feature request: Comply with CommonMark specification

### ISSUE THREAD

**Issue by getreu** [jugglerchris/rust-html2text#48]
Your library looks very promising. Unfortunately I can not use it because:

1. _html2text_'s output is not sufficiently _CommonMark_ compliant yet,
2. the HTML's metadata is not converted into a YAML metadata block (see pandoc's `yaml_metadata_block`) 

Let me explain my use case more in detail: I tested if I could replace in my toolchain:

    pandoc --standalone -f html -t markdown_strict+yaml_metadata_block+pipe_tables

with:

    html2text

This would allow me to do this:

    curl $(xclip -o)| thml2text | tp-note

and even integrate your library into [tp-note](https://github.com/getreu/tp-note). Then the above would look like this:

    curl $(xclip -o) | tp-note

[Tp-Note](https://github.com/getreu/tp-note) comes with a document viewer that renders the content with [pulldown-cmark](https://crates.io/crates/pulldown-cmark/0.8.0) which is compliant with the [CommonMark](https://commonmark.org/) specification.

As the de facto official specification for Markdown is _CommonMark_, making Html2text compatible with it, would open a wider range of use cases (mine included).
Another advantage: CommonMark has a [validation test suite](http://code.commonmark.org/).

What do you think?

**Comment by jugglerchris** [comment]
Hi,

As you've noticed the text output is currently inspired by markdown rather than trying to be actually valid markdown.

However I do like the idea of a full markdown output mode, and CommonMark would seem to be a good choice.  (I don't think I'd want markdown to be the default for the executable, but it'd be fine to add an option `html2text --md` or even a separate example program `html2md`).

**Comment by jugglerchris** [comment]
I had a look at the validation test suite - unfortunately I've found it's less useful for the reverse case than you'd think.  It has tests that some particular markdown gives the right HTML output - but there are multiple markdown inputs which could give that output - there's no one right answer.

Having thought about it (for quite a long time!), I think that strict markdown output is not an explicit goal of this project (though also not an anti-goal).

I'd be very happy to take changes which (probably optionally) make the output closer to, or conformant to CommonMark, but I don't consider it a bug that the output isn't markdown and don't have plans to work on it myself - so I'm going to close this issue with apologies.

If markdown is the only output you need, I suggest a crate which does have that goal (e.g. [html2md](https://crates.io/crates/html2md) - this is not a recommendation, I haven't used it).

**Comment by getreu** [comment]
Thank you for this long reply and for sharing your thoughts.

> but there are multiple markdown inputs which could give that output - there's no one right answer.

This is right. The specification establishes a set of functions

      HTML -> Markdown 
      n -> 1

This means to be (backwards) compliant, you can choose one particular `n` for each function.

Why does CommonMark matter? My use case is: [tpnote](https://github.com/getreu/tp-note) reads the HTML clipboard and creates with (some/your?) converter a Markdown note file. The same file is immediately rendered back into HTML for live previewing. This renderer is CommonMark compliant, which is the standard nowadays.

**Comment by jugglerchris** [comment]
Right, the `N -> 1` part means that the test suite is less helpful (you have to pick the "right" one for every test, even if other outputs would be correct in this case).

The discussion on #91 may be useful - I guess the Markdown decorator there gets most of the way, except for tables?

**Comment by getreu** [comment]
> Right, the N -> 1 part means that the test suite is less helpful (you have to pick the "right" one for every test, even if other outputs would be correct in this case).

We should not mix the specification and the Md->HTML test suite. N -> 1 means, the language specification allows more than one syntax to express the same semantic: e.g.:

```md
# Title
```

```md
Title
=====
```

has the same meaning. Whatever you prefer: being "Commonmark compliant" means, your output passes the test suite.

**Comment by jugglerchris** [comment]
Agreed - I was just saying that reversing the inputs and outputs from the test suite doesn't make it usable for testing the inverse mapping.  (Maybe something could be done with a forward-back double mapping)
